mockserver/main.go

```diff
@@ -0,0 +1,141 @@
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"io"
+	"log"
+	"net/http"
+	"path"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/gorilla/mux"
+)
+
+const (
+	HealthCheckMessage = "healthcheck"
+	SuccessMessage     = "success"
+)
+
+var (
+	CertFilePath = path.Join("certificates", "ssl", "certificate.crt")
+	KeyFilePath  = path.Join("certificates", "private.key")
+)
+
+type transactionHttpServer struct {
+	transactions uint32
+	startTime    time.Time
+}
+
+type TransactionPayload struct {
+	TransactionsPerMinute float64 `json:"GetNumberOfTransactionsPerMinute"`
+}
+
+func healthCheck(w http.ResponseWriter, _ *http.Request) {
+	if _, err := io.WriteString(w, HealthCheckMessage); err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		log.Printf("Unable to write response: %v", err)
+		return
+	}
+	w.WriteHeader(http.StatusOK)
+}
+
+func (ts *transactionHttpServer) checkTransactionCount(w http.ResponseWriter, _ *http.Request) {
+	var message string
+	var t = atomic.LoadUint32(&ts.transactions)
+	if t > 0 {
+		message = SuccessMessage
+	}
+	log.Printf("\033[31m Time: %d | checkTransactionCount msg: %s | %d\033[0m \n", time.Now().Unix(), message, t)
+	if _, err := io.WriteString(w, message); err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		io.WriteString(w, err.Error())
+		log.Printf("Unable to write response: %v", err)
+		return
+	}
+	w.WriteHeader(http.StatusOK)
+}
+
+func (ts *transactionHttpServer) acknowledgeTransaction(w http.ResponseWriter, _ *http.Request) {
+	atomic.AddUint32(&ts.transactions, 1)
+
+	// Built-in latency
+	log.Printf("\033[31m Time: %s | transaction received \033[0m \n", time.Now().String())
+	time.Sleep(15 * time.Millisecond)
+	w.WriteHeader(http.StatusOK)
+}
+
+// Retrieve number of transactions per minute
+func (ts *transactionHttpServer) GetNumberOfTransactionsPerMinute(w http.ResponseWriter, _ *http.Request) {
+	// Calculate duration in minutes
+	duration := time.Now().Sub(ts.startTime)
+	transactions := float64(atomic.LoadUint32(&ts.transactions))
+	tpm := transactions / duration.Minutes()
+
+	w.Header().Set("Content-Type", "application/json")
+	if err := json.NewEncoder(w).Encode(TransactionPayload{tpm}); err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		io.WriteString(w, err.Error())
+		log.Printf("Unable to write response: %v", err)
+	}
+}
+
+// Starts an HTTP server that receives request from validator only to verify the data ingestion
+func StartHttpServer() {
+	var wg sync.WaitGroup
+	log.Println("\033[31m Starting Server \033[0m")
+	store := transactionHttpServer{startTime: time.Now()}
+	//2 servers one for receiving the data , one for verify data
+	dataApp := mux.NewRouter()
+	dataReceiverServer := &http.Server{Addr: ":443", Handler: dataApp}
+	verificationRequestServer := http.NewServeMux()
+	appServer := &http.Server{Addr: ":8080", Handler: verificationRequestServer}
+	go func(ts *transactionHttpServer) {
+		wg.Add(1)
```

Comment:
Incrementing wait groups should be done outside of the go routine.

>Typically this means the calls to Add should execute before the statement creating the goroutine or other event to be waited for.

https://pkg.go.dev/sync#WaitGroup.Add